File: ruby/robot-name/robot_name.rb
module BookKeeping
  VERSION = 2
end

class Robot

  attr_reader :name
 @@used_names = []
  def initialize
    while true
      name = (letter_choose(2) + number_choose(3)).join
      unless @@used_names.include?(name)
        @name = name
        @@used_names << @name
        break
      end
    end
  end

  def reset
    initialize
  end

  def letter_choose(num)
    num.times.map { [*'A'..'Z'].sample }
  end

  def number_choose(num)
    num.times.map { (0..9).to_a.sample.to_s }
  end

end

File: python/robot-name/robot_name.py
import random
import string
historic_names = []
class Robot(object):
    
    def __init__(self):
        self.reset()

    def reset(self):
        self.name = ''.join(random.choice(string.ascii_letters).upper() for x in range(2)) + str(random.randrange(100, 999))
        while self.name in historic_names:
            self.name = ''.join(random.choice(string.ascii_letters).upper() for x in range(2)) + str(random.randrange(100, 999))
        historic_names.append(self.name)